### PR TITLE
Add act/Docker integration tests for subprocess boundary and _extract…

### LINKED ISSUE #39: Add Integration Test Coverage for Act Subprocess Boundary
### Problem

The eval (Test 31) identified a trust inversion at the core of the project: the priority queue, dependency resolver, image registry scoring algorithm, and orchestrator are all unit-tested through mocks, but the act subprocess boundary — where the tool actually delivers its value — has zero integration test coverage. Error extraction from act output uses keyword heuristics (`_extract_error()` in `cli/localci/core/executor.py`) that scan for "error:" and "failed" in captured text. These heuristics have never been validated against actual act output. The CI pipeline runs pytest on three Python versions but tests only the mocked path — no Docker or act is exercised.

### Acceptance Criteria

- [ ] Integration test suite created that invokes `localci run` against a minimal test workflow with actual act/Docker execution
- [ ] Test fixture: minimal GitHub Actions workflow file (e.g., `echo "Hello"` step) that can be run via act in Docker
- [ ] Test case: successful workflow execution — verify exit code, output capture, and result reporting
- [ ] Test case: failing workflow execution — verify `_extract_error()` correctly extracts the error from real act output
- [ ] Test case: invalid workflow file — verify graceful error handling (no raw traceback)
- [ ] Integration tests marked as optional in CI (require Docker/act) with a clear `pytest.mark.integration` marker
- [ ] CI workflow optionally runs integration tests when Docker is available (separate job or manual trigger)
- [ ] Tests pass in CI (if applicable)
- [ ] PR approved by at least 1 reviewer

### Implementation Notes

Create a `cli/tests/integration/` directory with `conftest.py` that checks for act and Docker availability and skips if missing. Use `pytest.mark.integration` so the tests are excluded by default (`-m "not integration"` in the default test command). The minimal test workflow should be in `cli/tests/fixtures/integration/` — a `.github/workflows/test.yml` with a single `ubuntu-latest` job and an `echo` step. For the failing workflow test, use a step like `run: exit 1` to produce a known error. Capture real act output and validate `_extract_error()` against it — this is the highest-value test since it validates the untested heuristic. Consider adding a GitHub Actions workflow job that installs act and Docker, runs the integration tests, and reports separately from the unit test job.

### References

- Eval finding: Test 31 (zero integration test coverage for act subprocess boundary)
- Related files: `cli/localci/core/executor.py` (`_extract_error()` method), `cli/tests/test_executor.py` (mocked tests), `cli/tests/fixtures/`, `.github/workflows/ci.yml`

### PR DESCRIPTION
**Pull request by henry0816191**
## Summary

Adds optional pytest integration coverage for the real act/Docker subprocess path (w4_issue_02 / Test 31). Unit tests mock act extensively; these tests validate `JobExecutor.run()`, `localci run`, and `_extract_error()` against live act output.

## Changes

- **Integration test package** (`cli/tests/integration/`)
  - `test_act_subprocess.py` — success run and failing run with `_extract_error()` checked against captured act stdout/stderr
  - `test_localci_run.py` — `localci run` for success, failure, and invalid workflow (no traceback on parse errors)
  - `conftest.py` — skip when act/Docker missing; pull/tag `catthehacker/ubuntu:act-24.04` for CLI runs
- **Fixtures** (`cli/tests/fixtures/integration/project/`)
  - Minimal workflows: `test.yml` (echo), `test-fail.yml` (exit 1), `invalid.yml` (malformed YAML)
  - Minimal `.localci.yml` (cache off, single parallel job)
- **Pytest config** (`cli/pyproject.toml`) — `integration` marker; `norecursedirs` excludes integration from default collection
- **CI** (`.github/workflows/ci.yml`)
  - Unit matrix: `pytest -m "not integration"`
  - New job **Integration (act + Docker)**: install act v0.2.76, pull runner image, run integration tests
- **Docs** (`cli/README.md`) — how to run unit vs integration tests

## Design notes

- Workflows use `strategy.matrix.include` (one entry) because `WorkflowAnalyzer` only produces runnable matrix entries that way.
- Executor tests map act’s image directly; CLI tests tag the derived `capy-ubuntu-latest-gcc15:latest` name so the orchestrator finds a local image.
- `test_run_invalid_workflow` does not require act/Docker (parse fails before execution).

## Test plan

- [x] `cd cli && pytest -m "not integration"` — unit tests pass
- [x] `pytest tests/integration -m integration` — invalid-workflow test passes without act; others skip or run when act/Docker available
- [x] CI: unit job (3.10–3.12) and **Integration (act + Docker)** job green on GHA

## Related

- close https://github.com/cppalliance/local-ci-test-system/issues/39

## Summary by CodeRabbit

* **Tests**
  * Added a comprehensive integration test suite exercising CLI runs and subprocess execution (success, failure, invalid workflow), plus fixtures and session-level setup that require containerized execution.

* **Documentation**
  * Updated development testing instructions with explicit unit, coverage (excluding integration), and a separate integration test command.

* **Chores**
  * Split CI into unit and integration runs, added a dedicated integration job with a longer timeout and pinned workflow tooling; adjusted test commands to exclude integration tests from unit runs.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cppalliance/local-ci-test-system/pull/43?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by henry0816191** [comment]
@coderabbitai continue to review

**Review by bradjin8 — COMMENTED** [review]
Strong addition for #39 — real act/Docker coverage and a high-value `_extract_error` check in `test_act_subprocess.py`. CI layout (unit vs integration job, pinned act binary, `needs: [test]`) looks good.

Please address the inline comments (mainly docker timeout handling and end-to-end failure assertions on the `localci run` path) before I approve. Happy to re-review after updates.

File: cli/tests/integration/conftest.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    if not _act_available():
+        pytest.skip("act is not installed")
+    if not _docker_available():
+        pytest.skip("Docker is not available")
+
+
+@pytest.fixture(scope="session")
+def act_runner_image(require_act_and_docker: None) -> str:
+    """Pull the act runner image used for ubuntu-latest jobs."""
+    pull = subprocess.run(
```
Comment (bradjin8) [inline cli/tests/integration/conftest.py]: `subprocess.run(..., timeout=DOCKER_PULL_TIMEOUT)` raises `subprocess.TimeoutExpired` on slow networks; that will **fail** the session instead of skipping integration tests.

Please wrap the pull in `try/except subprocess.TimeoutExpired` and `pytest.skip(...)` (keep the existing non-zero `returncode` handling).

File: cli/tests/integration/conftest.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+def capy_image_tag(act_runner_image: str, require_act_and_docker: None) -> str:
+    """Tag the act runner image as the derived capy name for localci run."""
+    from localci.core.workflow import WorkflowAnalyzer
+
+    workflow_path = FIXTURE_PROJECT / ".github/workflows/test.yml"
+    entry = WorkflowAnalyzer().analyze(workflow_path).jobs[INTEGRATION_JOB_ID].matrix[0]
+    tag = derive_image_tag(entry)
+    assert tag is not None
+
+    tag_result = subprocess.run(
```
Comment (bradjin8) [inline cli/tests/integration/conftest.py]: Same as the docker pull above: `docker tag` uses a 60s timeout but only checks `returncode`. Catch `subprocess.TimeoutExpired` and skip with a clear message so a slow daemon doesn't fail the whole integration session.

File: cli/tests/integration/test_localci_run.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+@pytest.mark.usefixtures("capy_image_tag")
+def test_run_failure(
+    integration_project: tuple[Path, Path],
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    project, _logs_dir = integration_project
+    monkeypatch.chdir(project)
+    result = _run_localci(project, ".github/workflows/test-fail.yml")
+
+    assert result.exit_code == 1, result.output
```
Comment (bradjin8) [inline cli/tests/integration/test_localci_run.py:91]: Issue #39 calls out validating `_extract_error()` against **real** act output. `test_failing_job_extract_error` does that well at the executor layer; this CLI test only checks generic keywords in stdout.

Consider also loading `last-run.json` (or the job log under `logs_dir`) and asserting the stored failure `error_message` matches `JobExecutor._extract_error(captured_output)` so the full `localci run` path is covered end-to-end.

File: cli/tests/integration/conftest.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        )
+    return ACT_RUNNER_IMAGE
+
+
+@pytest.fixture(scope="session")
+def capy_image_tag(act_runner_image: str, require_act_and_docker: None) -> str:
+    """Tag the act runner image as the derived capy name for localci run."""
+    from localci.core.workflow import WorkflowAnalyzer
+
+    workflow_path = FIXTURE_PROJECT / ".github/workflows/test.yml"
```
Comment (bradjin8) [inline cli/tests/integration/conftest.py:76]: Nit: `capy_image_tag` derives the image name from `test.yml` only. That works today because `test-fail.yml` uses the same matrix shape, but if the failure fixture diverges later, `test_run_failure` could break silently. A short comment here (or derive the tag from the workflow under test) would make the coupling explicit.

File: cli/tests/integration/test_act_subprocess.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    )
+    return result, entry
+
+
+def test_successful_job_execution(
+    integration_project: tuple[Path, Path],
+    act_runner_image: str,
+    tmp_path: Path,
+) -> None:
+    project, _ = integration_project
```
Comment (bradjin8) [inline cli/tests/integration/test_act_subprocess.py]: Nit: `integration_project` already returns `(project, logs_dir)` and configures `.localci.yml` to use `tmp_path / "logs"`. Consider `project, logs_dir = integration_project` instead of recomputing `tmp_path / "logs"` — same path today, clearer intent.

File: cli/pyproject.toml [hunk 6]
```diff
@@ -31,3 +31,7 @@ packages = ["localci"]
 [tool.pytest.ini_options]
 testpaths = ["tests"]
 addopts = "-v --tb=short"
+markers = [
+    "integration: requires act and Docker; runs real subprocesses",
+]
+norecursedirs = ["integration"]
```
Comment (bradjin8) [inline cli/pyproject.toml:37]: Nit: `norecursedirs = ["integration"]` means `pytest -m integration` **without** `tests/integration/` collects zero tests. README documents the explicit path (good). Optional: add `-m "not integration"` to `addopts` so unit runs stay safe if integration tests ever move out of that directory name.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -20,15 +20,58 @@ jobs:
...
+          # From https://github.com/nektos/act/releases/download/v0.2.76/checksums.txt
+          ACT_LINUX_X86_64_SHA256: d4720e05e73ce634239fc0e7fa6b552f9ac79b4e6c878a8756bbfacce3c56720
+        run: |
+          curl -sSfL -o /tmp/act.tar.gz \
+            "https://github.com/nektos/act/releases/download/${ACT_VERSION}/act_Linux_x86_64.tar.gz"
+          echo "${ACT_LINUX_X86_64_SHA256}  /tmp/act.tar.gz" | sha256sum --check
+          sudo tar xzf /tmp/act.tar.gz -C /usr/local/bin act
+          act --version
+
+      - name: Pull act runner image
```
Comment (bradjin8) [inline .github/workflows/ci.yml]: Nit: the runner image is pulled here and again in session fixture `act_runner_image`. Harmless but redundant (~CI time). You could drop this step and rely on conftest, or skip the fixture pull when `CI=true` and the image is already present.

File: cli/README.md [hunk 8]
```diff
@@ -105,11 +105,14 @@ cli/
 # Editable install with dev dependencies
 pip install -e ".[dev]"
 
-# Run tests
+# Run unit tests (default; integration tests are excluded)
 pytest
```
Comment (bradjin8) [inline cli/README.md:111]: Nit: default `pytest` excludes integration via `norecursedirs`, while CI uses `-m "not integration"`. Consider noting both in the README (e.g. that plain `pytest` won't run integration tests, and coverage should use `-m "not integration"` as documented on the next line) so behavior stays obvious if layout changes.

File: cli/tests/integration/test_localci_run.py (hunk 3 above)
Comment (henry0816191) [reply cli/tests/integration/test_localci_run.py:91]: test_run_failure loads last-run.json, reads the job log (or CLI output as fallback), and asserts error_message in JobExecutor._extract_error(captured).

File: cli/tests/integration/conftest.py (hunk 4 above)
Comment (henry0816191) [reply cli/tests/integration/conftest.py:76]: A comment documents that the tag comes from test.yml and that test-fail.yml currently shares the same matrix shape.

File: cli/pyproject.toml (hunk 6 above)
Comment (henry0816191) [reply cli/pyproject.toml:37]: Intentionally not done — it would break pytest tests/integration -m integration in CI; norecursedirs plus README documentation cover unit-test exclusion instead.

File: cli/README.md (hunk 8 above)
Comment (henry0816191) [reply cli/README.md:111]: The Development section explains norecursedirs, -m "not integration" for coverage/CI, and the path + marker needed to run integration tests.

**Review by bradjin8 — APPROVED** [review]

**Review by wpak-ai — APPROVED** [review]
